tests/cpp/test_cache.cc

```diff
@@ -52,4 +54,63 @@ TEST(DMatrixCache, Basic) {
     }
   }
 }
+
+TEST(DMatrixCache, MultiThread) {
```

Comment:
I have run this test a few thousand times using the `gtest_repeat`, along with thread sanitizer.